### PR TITLE
Fix #7986: Warn on cyclic references when looking for extension methods

### LINKED ISSUE #7986: Extension methods do not see each other when overloaded
## minimized code

```Scala
case class Project(name: String)
def (name: String) dependencies = ???
def (project: Project) dependencies = project.name.dependencies
```

<details>
<summary>Compilation output</summary>

```scala
-- [E008] Member Not Found Error: /Users/kmetiuk/Projects/scala3/pg/sandbox/iss1.scala:3:51
3 |def (project: Project) dependencies = project.name.dependencies
  |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^
  |                              value dependencies is not a member of String
1 error found
```
</details>

## expectation
The following compiles fine (`dependencies` -> `dependenciez`):

```scala
case class Project(name: String)
def (name: String) dependencies = ???
def (project: Project) dependenciez = project.name.dependencies
```

So the original code should also compile.

Possibly related: #7375, #7401.

### PR DESCRIPTION
**Pull request by odersky**
Propagate cyclic reference errors when searching extension methods

### CODE REVIEW

**Comment by odersky** [comment]
The first attempt to simply propagate cyclic references breaks existing tests. We now issue a warning instead. On i7986.scala you get:
```
-- Warning: i7986.scala:3:46 ---------------------------------------------------
3 |def (project: Project) dependencies = project.name.dependencies // error, following a cyclic reference warning
  |                                      ^^^^^^^^^^^^
  |          An extension method was tried but could not be fully constructed
  |          since there was a cyclic reference involving method dependencies
-- [E008] Member Not Found Error: i7986.scala:3:51 -----------------------------
3 |def (project: Project) dependencies = project.name.dependencies // error, following a cyclic reference warning
  |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^
  |                              value dependencies is not a member of String
1 warning found
1 error found
```
Unfortunately I cannot test that a warning is issued with a .check file since the .check files did not show the warning (maybe it was suppressed since there was an error?).

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -3048,16 +3048,24 @@ class Typer extends Namer
...
               findRef(name, WildcardType, ExtensionMethod, tree.posd) match {
                 case ref: TermRef =>
                   extMethodApply(untpd.ref(ref).withSpan(tree.span), tree, mbrType)
                 case _ => EmptyTree
               }
-            catch {
-              case ex: TypeError => errorTree(tree, ex, tree.sourcePos)
-            }
+            catch
+              case ex: CyclicReference =>
```
Comment (anatoliykmetyuk) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: You could probably do:

```scala
              case ex: CyclicReference =>
                if ex.denot.is(Extension) && ex.denot.name == name
                  throw ex
                else
                  errorTree(tree, ex, tree.sourcePos)
```

It fixes the failing tests.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -3048,16 +3048,24 @@ class Typer extends Namer
...
+                if ex.denot.is(Extension) && ex.denot.name == name then
+                  origCtx.warning(
+                    em"""An extension method was tried but could not be fully constructed
+                        |since there was a cyclic reference involving ${ex.denot.symbol.showLocated}""",
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: It turns out to be quite hard to figure out accurately that a return type is missing.

**Comment by odersky** [comment]
There was something really funky going on. Basically, we had

    extension (sym: Symbol)
      def documentation = sym.documentation match ...

How is this not an infinite loop? Well, name resolution tried the current method,
failed because of a cyclic reference and then fell back to an alternative
extension method elsewhere. So compilation relied crucially on a missing result type
in `documentation`. This is terrible! This PR plugs the specific hole, but
I fear there are others that could be exploited in user code.

Morale: Leaving out result types has an upside in that it detects compiler
bugs. But for robust software, it's better do add return types to all public methods.

**Comment by odersky** [comment]
@romanowski Can you check that my changes to BasicSupport are correct?

**Review by romanowski — APPROVED** [review]
Changes in scal3doc looks ok. I've take a look why we had `documentation` calling `documentation` and it seems that originally we used `comment` method that gets renamed later on and we simply changed it and since it worked we leaved it be wihtout realizing that we exploited the compiler.

**Review by anatoliykmetyuk — APPROVED** [review]

**Comment by odersky** [comment]
#10902 provides a better fix

**Pull request closed without merging**